library/src/scala/quoted/compiletime/Constant.scala

```diff
@@ -0,0 +1,273 @@
+package scala.quoted.compiletime
+
+/////// Constant ///////////////////////////////////////////////////////////////
+
+sealed trait Constant {
+
+  def value: Any
+
+  final def show(using p: Printer[Constant]): String = p.show(this)
+
+}
+object Constant {
+
+  def quoted(using quotes: Quotes): Constant.Module = quotes.reflectV2.Constant
+  given moduleConversion: (quotes: Quotes) => Conversion[Constant.type, Constant.Module] = _ => quotes.reflectV2.Constant
+
+  def unapply(x: Constant): Some[Any] = Some(x.value)
+
+  trait Module private[compiletime] () {}
+
+}
+
+/////// BooleanConstant ///////////////////////////////////////////////////////////////
+
+trait BooleanConstant private[compiletime] () extends Constant {
+
+  override def value: Boolean
+
+}
+object BooleanConstant {
+
+  def quoted(using quotes: Quotes): BooleanConstant.Module = quotes.reflectV2.BooleanConstant
+  given moduleConversion: (quotes: Quotes) => Conversion[BooleanConstant.type, BooleanConstant.Module] = _ => quotes.reflectV2.BooleanConstant
+
+  def unapply(x: BooleanConstant): Some[Boolean] = Some(x.value)
+
+  trait Module private[compiletime] () {
+    def apply(x: Boolean): BooleanConstant
+    def make(x: Boolean): BooleanConstant
+  }
+
+}
```

Comment:
public definition of `BooleanConstant` + `BooleanConstant.Module`

as long as you have a `given Quotes`, can be used at the top level as such:
```scala
def myFunction(bool: Boolean)(using Quotes): Nothing = {
  val ex1: BooleanConstant = BooleanConstant.quoted.make(bool)
  val ex2: Constant = BooleanConstant.make(bool)
  ???
}
```
Thanks to the `Conversion`, we can treat `object BooleanConstant` exactly like we would expect to treat a normal companion object, but we still achieve the desired obfuscation of the private API.